src/librustc/sir.rs

```diff
@@ -0,0 +1,136 @@
+//! Serialised Intermediate Representation (SIR).
+//!
+//! SIR is built in-memory during LLVM code-generation, and finally placed into a dedicated ELF
+//! section at link time.
+
+use std::default::Default;
+use std::io::{self, Write};
+use rustc_index::{newtype_index, vec::{Idx, IndexVec}};
+use rustc_data_structures::fx::FxHashMap;
+use ykpack;
+
+// Duplicates of LLVM types defined elsewhere, copied to avoid cyclic dependencies. Whereas the
+// LLVM backend expresses pointers to these using references, we use raw pointers so to as avoid
+// introducing lifetime parameters to the SirCx (and thus into TyCtxt and every place that uses
+// it).
+extern { pub type Value; }
+extern { pub type BasicBlock; }
+
+newtype_index! {
+    pub struct SirFuncIdx {
+        DEBUG_FORMAT = "SirFuncIdx({})"
+    }
+}
+
+// The index of a block within a function.
+// Note that these indices are not globally unique. For a globally unique block identifier, a
+// (SirFuncIdx, SirBlockIdx) pair must be used.
+newtype_index! {
+    pub struct SirBlockIdx {
+        DEBUG_FORMAT = "SirBlockIdx({})"
+    }
+}
+
+/// Sir equivalents of LLVM values.
+#[derive(Debug)]
+pub enum SirValue {
+    Func(SirFuncIdx),
+}
+
+impl SirValue {
+    pub fn func_idx(&self) -> SirFuncIdx {
+        let Self::Func(idx) = self;
+        *idx
+    }
+}
+
+pub struct SirCx {
+    /// Maps an opaque LLVM `Value` to its SIR equivalent.
+    pub llvm_values: FxHashMap<*const Value, SirValue>,
+    /// Maps an opaque LLVM `BasicBlock` to the function and block index of it's SIR equivalent.
+    pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
+    /// Function store. Also owns the blocks
+    pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+}
+
+impl SirCx {
+    pub fn new() -> Self {
+        Self {
+            llvm_values: Default::default(),
+            llvm_blocks: FxHashMap::default(),
+            funcs: Default::default(),
+        }
+    }
+
+    pub fn add_func(&mut self, value: *const Value, symbol_name: String) {
+        let idx = SirFuncIdx::from_usize(self.funcs.len());
+
+        self.funcs.push(ykpack::Body{
+            symbol_name,
+            blocks: Default::default(),
+            num_locals: 0,  // FIXME
```

Comment:
Can we make these `assert`s on something? I always find `FIXME` dangerous, but I know it's occasionally unavoidable.

Reply:
I'm not sure what the assertion would be or if we even need these fields in the new SIR. I only included them so as to minimise changes to ykpack.

Reply:
If we don't know if we need them, wouldn't it be better to bite the pullet re: ykpack? Otehrwise I can imagine no-one will ever be brave enough to remove them!

Reply:
I suppose it doesn't hurt. We can always bring it back later.